Clean up auto yaw behaviour

peterbarker opened the pull request:
~~This is based on the `ok_to_enter()` PR, https://github.com/ArduPilot/ardupilot/pull/7402 - that one should be merged first.~~

amilcarlucas commented:
I have a very old PR that also touches this code : https://github.com/ArduPilot/ardupilot/pull/5507
 I hope that one gets merged first :)

peterbarker commented:
@amilcarlucas The one substantial change in your PR is something that stood out to be as being a bit odd!

That block of code seems to appear in two places, three with your patch.

peterbarker commented:
On Tue, 9 Jan 2018, Francisco Ferreira wrote:

> It would be nice to change all _AutoYaw definitions to a single file,
> either a new one or mode.cpp; having them all over the place is quite bad.

Yep, as soon as you're happy with them all I'll add a patch on top 
concentrating with (autoyaw.cpp)

> 
> +        // auto flight mode's yaw mode
> +        uint8_t _mode = AUTO_YAW_LOOK_AT_NEXT_WP;
> 
> Shouldn't this be of type autopilot_yaw_mode?

I figured we could save 3 bytes of memory here...

>      case AUTO_YAW_ROI:
>          // point towards a location held in yaw_look_at_WP
> 
> This comment (which is also in the _roi_yaw declaration) seems a bit off
> since there isn't (nor was) a yaw_look_at_WP variable.

Will fix.

> >      }
>  }
> 
> -// returns yaw rate held in auto_yaw_rate and normally set by SET_POSITION
> _TARGET mavlink messages (positive it clockwise, negative is counter clockw
> ise)
> -float Copter::get_auto_yaw_rate_cds(void)
> +// returns yaw rate normally set by SET_POSITION_TARGET mavlink
> +// messages (positive it clockwise, negative is counter clockwise)
> 
> Since we are changing: positive it -> positive is

Fixed.

Changes fixup'd in and pushed up.

OXINARF commented on ArduCopter/mode.h:
```diff
@@ -26,13 +26,70 @@ class Mode {
+        // Yaw will point at this location if mode is set to AUTO_YAW_ROI
+        Vector3f roi_WP;
+
+        // bearing from current location to the yaw_look_at_WP
```
Missed it here.

peterbarker commented:
On Wed, 10 Jan 2018, Francisco Ferreira wrote:

> Alright, I don't care much, although I prefer types to be more explicit.

Yep, if we didn't have a getter I'd probably go with the enum.

> +        // bearing from current location to the yaw_look_at_WP
> 
> Missed it here.

Fixed.

I've also created autoyaw.cpp - and run astyle over it :-)

peterbarker commented:
I've moved this off the top of ok-to-enter.  I'd still prefer that one go in first, but if that one doesn't maybe this one might...

OXINARF commented on ArduCopter/mode_auto.cpp:
```diff
@@ -643,66 +650,65 @@ void Copter::set_auto_yaw_mode(uint8_t yaw_mode)
     if (roi_location.alt == 0 && roi_location.lat == 0 && roi_location.lng == 0) {
         // set auto yaw mode back to default assuming the active command is a waypoint command.  A more sophisticated method is required to ensure we return to the proper yaw control for the active command
-        set_auto_yaw_mode(get_default_auto_yaw_mode(false));
+        auto_yaw.set_mode_to_default(false);
```
The auto_yaw isn't needed right? This method belongs to the class now. There are others like this below.

peterbarker commented:
On Mon, 19 Mar 2018, Francisco Ferreira wrote:

> > @@ -15,7 +15,7 @@ enum autopilot_yaw_mode {
>      AUTO_YAW_HOLD =             0,  // pilot controls the heading
> 
> I wonder if we should move this enum to the _AutoYaw class. The disadvantage
> is that we would have to write Mode::_AutoYaw::FIXED where it is used, which
> is a bit bigger.

... and ugly :-)  Maybe later...

> +    // Navigation Yaw control
> +    class _AutoYaw {
> 
> Not blocking this PR, but I'm feeling more that the underscore should go
> away, it's an inner class but it's public.

Yep, this is a discussion I wanted to have.  I really don't want them to 
look like an actual flightmode, thus the underscore.  "Mode::Stabilize" vs 
"Mode::AutoYaw" == confusing in my mind.

> +        uint8_t roi_yaw_counter = 0;
> Initialization shouldn't be needed, right?

Removed.

> +        fixed_yaw_slewrate = AUTO_YAW_SLEW_RATE;
> _fixed...

Fixed.

> -        set_auto_yaw_mode(get_default_auto_yaw_mode(false));
> +        auto_yaw.set_mode_to_default(false);
> 
> The auto_yaw isn't needed right? This method belongs to the class now. There
> are others like this below.

Fixed.

Pushed these changes up.

OXINARF commented:
> Yep, this is a discussion I wanted to have.  I really don't want them to
look like an actual flightmode, thus the underscore.  "Mode::Stabilize" vs
"Mode::AutoYaw" == confusing in my mind.

It's Copter::ModeStabilize vs Copter::Mode::AutoYaw. I get that it might not make much difference visually, but it is different, one is an inner class of Copter, the other is an inner class of Mode (which is an inner class of Copter). I prefer it without underscore, but, as I said, not blocking this PR on it.

I'll merge this soon, unless someone has anything against it.

amilcarlucas commented on ArduCopter/autoyaw.cpp:
```diff
@@ -0,0 +1,213 @@
+float Copter::Mode::_AutoYaw::roi_yaw()
+{
+    roi_yaw_counter++;
+    if (roi_yaw_counter >= 4) {
```
The idea here is to reduce the rate to 100 Hz, but it is hardcoded and will give different results if the loop speed changes.
So how about calculating the decimation value once on initialization and then keep the result in a private variable (_roi_yaw_update_decimation = LOOP_RATE_PARAMETER / 100; )

amilcarlucas replied:
Ah, and add a comment here explaining that this is to reduce the update rate to 100Hz

amilcarlucas commented on ArduCopter/commands_logic.cpp:
```diff
@@ -1084,13 +1084,13 @@ bool Copter::ModeAuto::verify_within_distance()
+    auto_yaw.set_mode(AUTO_YAW_FIXED);
 
     // check if we are within 2 degrees of the target heading
-    if (labs(wrap_180_cd(ahrs.yaw_sensor-copter.yaw_look_at_heading)) <= 200) {
```
On my SITL simulations, my YAW error is always around 2 deg. Maybe we should make this 200 a parameter, so that users that require more precision can change it ?

amilcarlucas commented on ArduCopter/mode_guided.cpp:
```diff
@@ -662,9 +662,9 @@ void Copter::ModeGuided::set_desired_velocity_with_accel_and_fence_limits(const
 {
     if (use_yaw) {
-        set_auto_yaw_look_at_heading(yaw_cd / 100.0f, 0.0f, 0, relative_angle);
+        auto_yaw.set_fixed_yaw(yaw_cd / 100.0f, 0.0f, 0, relative_angle);
```
How about replacing this with a multiplication ?

amilcarlucas commented on ArduCopter/mode_rtl.cpp:
```diff
@@ -215,17 +215,17 @@ void Copter::ModeRTL::loiterathome_run()
-        if (copter.auto_yaw_mode == AUTO_YAW_RESETTOARMEDYAW) {
+        if (auto_yaw.mode() == AUTO_YAW_RESETTOARMEDYAW) {
             // check if heading is within 2 degrees of heading when vehicle was armed
             if (labs(wrap_180_cd(ahrs.yaw_sensor-copter.initial_armed_bearing)) <= 200) {
```
Again an hardcoded 2 deg. IMHO this should at least be moved to a define. If it is used more than once.

peterbarker commented:
On Tue, 20 Mar 2018, Amilcar Lucas wrote:

> +                copter.flightmode->auto_yaw.set_fixed_yaw(
> +                    (float)packet.param3 / 100.0f,
> 
> How about replacing this with a multiplication ?

That would be a good change, but I'm really trying to just reorganise code 
here :-)  Maybe another patch on top while we've lost history on the line 
anyway?

> +        // TO-DO: expand handling of the do_nav_roi to support all modes of
>  the MAVLink.  Currently we only handle mode 4 (see below)
> Should this be mode 3 instead ?

Possibly.  I just naturally assume comments are wrong.  Again, just moving 
stuff, not trying to correct it :-)

> +// roi_yaw - returns heading towards location held in roi
> +float Copter::Mode::_AutoYaw::roi_yaw()
> +{
> +    roi_yaw_counter++;
> +    if (roi_yaw_counter >= 4) {
> 
> The idea here is to reduce the rate to 100 Hz, but it is hardcoded and will
> give different results if the loop speed changes.
> So how about calculating the decimation value once on initialization and
> then keep the result in a private variable (_roi_yaw_update_decimation =
> LOOP_RATE_PARAMETER / 100; )

That's going to be a very worthwhile change.  I'm quite happy to do it, 
but as a separate PR.  Many other places we'll keep around a millis() 
rather than decimation - why do you think we should decimate here?

>      // check if we are within 2 degrees of the target heading
> -    if (labs(wrap_180_cd(ahrs.yaw_sensor-copter.yaw_look_at_heading)) <= 20
> 0) {
> 
> On my SITL simulations, my YAW error is always around 2 deg. Maybe we should
> make this 200 a parameter, so that users that require more precision can
> change it ?

Sounds reasonable.

> +// roi_yaw - returns heading towards location held in roi
> +float Copter::Mode::_AutoYaw::roi_yaw()
> +{
> +    roi_yaw_counter++;
> +    if (roi_yaw_counter >= 4) {
> 
> Ah, and add a comment here explaining that this is to reduce the update rate
> to 100Hz

Just moving code, not fixing it :-)

>              // check if heading is within 2 degrees of heading when vehicle
>  was armed
>              if (labs(wrap_180_cd(ahrs.yaw_sensor-copter.initial_armed_beari
> ng)) <= 200) {
> 
> Again an hardcoded 2 deg. IMHO this should at least be moved to a define. If
> it is used more than once.

I like the parameter idea better :-)  I'm not sure where we would hang 
that parameter, however, so at least a define.....


I'd really prefer not to scope-creep this PR.  You have raised some very, 
very good points here, but they are changes better suited to a different 
PR IMO.

amilcarlucas commented:
Needs a huge rebase :(

peterbarker commented:
@amilcarlucas Rebased.  That was unpleasant.

amilcarlucas commented:
Can this one go in now ?

rmackay9 commented:
TBH, I'm not a huge fan.  I don't really see the point of adding another class into the mix... and why the underscore before the class name.  I don't think we've done that before..

OXINARF commented:
@rmackay9 
I like the class as it clearly separates things. One of my biggest troubles with vehicle code until today is finding where stuff is or what it respects too. Sometimes variable names make it clear, but sometimes it doesn't, especially when it would make the name very long.

Regarding the underscore: as I've said above, I've come to think that it shouldn't have the underscore too. But @peterbarker's thinking was that it is an inner class and that's a common scheme to use when a inner class is private - it isn't the case here though, it's an inner class, but it's a public one. @peterbarker also wanted to make clear it isn't a mode (although I think that's different already: `Copter::ModeAuto` vs `Copter::Mode::AutoYaw`).

If it wasn't clear, for me, this is ready to go.

peterbarker commented:
Removed the underscore.  Now just AutoYaw

amilcarlucas commented:
Needs a rebase now. Once this one goes in I'll do a PR on top of it.

rmackay9 commented:
Ok, I'm convinced it is an improvement.  I especially like that the class was moved to a new file instead of being inside mode.cpp which is really getting big.
